impl/container/src/main/java/io/serverlessworkflow/impl/container/executors/ContainerRunner.java

```diff
@@ -21,115 +21,122 @@
+import java.io.IOException;
 import java.time.Duration;
 import java.util.ArrayList;
-import java.util.List;
+import java.util.Collection;
+import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.TimeUnit;
-import java.util.function.Function;
 
 class ContainerRunner {
 
   private static final DefaultDockerClientConfig DEFAULT_CONFIG =
       DefaultDockerClientConfig.createDefaultConfigBuilder().build();
 
-  private static final DockerClient dockerClient =
-      DockerClientImpl.getInstance(
-          DEFAULT_CONFIG,
-          new ApacheDockerHttpClient.Builder().dockerHost(DEFAULT_CONFIG.getDockerHost()).build());
+  private static class DockerClientHolder {
+    private static final DockerClient dockerClient =
+        DockerClientImpl.getInstance(
+            DEFAULT_CONFIG,
+            new ApacheDockerHttpClient.Builder()
+                .dockerHost(DEFAULT_CONFIG.getDockerHost())
+                .build());
+  }
 
-  private final CreateContainerCmd createContainerCmd;
-  private final Container container;
-  private final List<ContainerPropertySetter> propertySetters;
-  private final WorkflowDefinition definition;
+  private final Collection<ContainerPropertySetter> propertySetters;
+  private final Optional<WorkflowValueResolver<Duration>> timeout;
+  private final ContainerCleanupPolicy policy;
+  private final String containerImage;
 
-  private ContainerRunner(
-      CreateContainerCmd createContainerCmd, WorkflowDefinition definition, Container container) {
-    this.createContainerCmd = createContainerCmd;
-    this.definition = definition;
-    this.container = container;
-    this.propertySetters = new ArrayList<>();
+  private ContainerRunner(ContainerRunnerBuilder builder) {
+    this.propertySetters = builder.propertySetters;
+    this.timeout = Optional.ofNullable(builder.timeout);
+    this.policy = builder.policy;
+    this.containerImage = builder.containerImage;
   }
 
   CompletableFuture<WorkflowModel> start(
       WorkflowContext workflowContext, TaskContext taskContext, WorkflowModel input) {
     return CompletableFuture.supplyAsync(
         () -> startSync(workflowContext, taskContext, input),
-        definition.application().executorService());
+        workflowContext.definition().application().executorService());
   }
 
   private WorkflowModel startSync(
       WorkflowContext workflowContext, TaskContext taskContext, WorkflowModel input) {
-    try {
-      var resolver = new StringExpressionResolver(workflowContext, taskContext, input);
-      applyPropertySetters(resolver);
-      pullImageIfNeeded(container.getImage());
-
-      String id = createAndStartContainer();
-      int exit = waitAccordingToLifetime(id, workflowContext, taskContext, input);
-      if (exit == 0) {
-        return input;
-      }
+    Integer exit = executeContainer(workflowContext, taskContext, input);
+    if (exit == null || exit == 0) {
```

Comment:
Its unclear to me if exit == null is equivalent to 0, but I respected the previous implementation